File: packages/optimizer/src/lbfgs.rs
//! This module implements an L-BFGS optimizer using techniques from the following three books:
//!
//! - _Convex Optimization_ by Boyd and Vandenberghe, 2009 edition
//! - _Engineering Optimization_ by Rao, 2009 edition
//! - _Numerical Optimization_ by Nocedal and Wright, 1999 edition
//!
//! A few functions in this module take a parameter of type `impl FnMut(&[f64], &mut [f64]) -> f64`.
//! The first parameter is the point at which to evaluate the objective function, and the second
//! parameter is an output parameter to hold the gradient at that point. The function should store
//! the gradient in that output parameter and then return the objective value.

/// Configuration options for L-BFGS.
#[derive(Clone, Copy, Debug)]
pub struct Config {
    /// The number of vector pairs to store for L-BFGS. See page 224 of Nocedal and Wright.
    pub m: usize,

    /// Constant for the Armijo condition. See page 37 of Nocedal and Wright.
    pub armijo: f64,

    /// Constant for the Wolfe condition. See page 39 of Nocedal and Wright.
    pub wolfe: f64,

    /// The minimum interval size for line search.
    pub min_interval: f64,

    /// The maximum number of steps for line search.
    pub max_steps: usize,

    /// A small positive constant to add to a denominator that might be zero.
    pub epsd: f64,
}

/// All L-BFGS state that needs to be kept between iterations, other than the current point.
#[derive(Clone, Debug)]
pub struct State {
    /// The previous point.
    pub x: Vec<f64>,

    /// The previous gradient.
    pub grad: Vec<f64>,

    /// See page 224 of Nocedal and Wright.
    pub s_y: Vec<(Vec<f64>, Vec<f64>)>,
}

/// Return the dot product of `u` and `v`.
fn dot(u: &[f64], v: &[f64]) -> f64 {
    u.iter().zip(v).map(|(a, b)| a * b).sum()
}

/// Return the line search step size, having set `x` to the new point.
///
/// - `x0` is the current point.
/// - `r` is the descent direction, preconditioned by L-BFGS.
/// - `fx0` is the objective at `x0`.
/// - `grad` is the gradient at `x0`, and is then used as scratch space; don't depend on its value.
/// - `x` will hold the new point.
fn line_search(
    cfg: Config,
    mut f: impl FnMut(&[f64], &mut [f64]) -> f64,
    x0: &[f64],
    r: &[f64],
    fx0: f64,
    grad: &mut [f64],
    x: &mut [f64],
) -> f64 {
    let n = x.len();

    let duf_at_x0 = -dot(r, grad);

    let mut a = 0.;
    let mut b = f64::INFINITY;
    let mut t = 1.;
    let mut j = 0;

    loop {
        for i in 0..n {
            x[i] = x0[i] - t * r[i];
        }

        if (a - b).abs() < cfg.min_interval || j > cfg.max_steps {
            break;
        }

        let fx = f(x, grad);
        let is_armijo = fx <= fx0 + cfg.armijo * t * duf_at_x0;
        let is_wolfe = -dot(r, grad) >= cfg.wolfe * duf_at_x0; // weak Wolfe condition

        if !is_armijo {
            b = t;
        } else if !is_wolfe {
            a = t;
        } else {
            break; // found good interval
        }

        if b < f64::INFINITY {
            t = (a + b) / 2.; // already found Armijo
        } else {
            t = 2. * a; // did not find Armijo
        }

        j += 1;
    }

    t
}

/// Perform the first step of L-BFGS at point `x`, updating it and returning the initial `State`.
pub fn first_step(
    cfg: Config,
    mut f: impl FnMut(&[f64], &mut [f64]) -> f64,
    x: &mut [f64],
) -> State {
    let n = x.len();
    let x0 = x.to_vec();

    let mut grad = vec![0.; n];
    let fx = f(x, &mut grad);

    let r = grad.clone();
    line_search(cfg, f, &x0, &r, fx, &mut grad, x);

    State {
        x: x0,
        grad: r,
        s_y: vec![],
    }
}

/// Information after a step of L-BFGS.
#[derive(Clone, Copy, Debug)]
pub struct Info<'a> {
    /// Data about previous steps.
    pub state: &'a State,

    /// The objective value at the current point.
    pub fx: f64,

    /// The preconditioned descent direction.
    pub r: &'a [f64],

    /// The current point.
    pub x: &'a [f64],

    /// The line search step size.
    pub t: f64,
}

/// Perform L-BFGS steps on `x`, returning once `stop` returns `Some`.
pub fn step_until<T>(
    cfg: Config,
    mut f: impl FnMut(&[f64], &mut [f64]) -> f64,
    x: &mut [f64],
    state: &mut State,
    mut stop: impl FnMut(Info) -> Option<T>,
) -> T {
    let n = x.len();
    let mut grad = vec![0.; n];

    let mut rho = vec![0.; cfg.m];
    let mut alpha = vec![0.; cfg.m];
    let mut q = vec![0.; n];
    let mut r = vec![0.; n];

    loop {
        let fx = f(x, &mut grad);

        if state.s_y.len() < cfg.m {
            state.s_y.push((
                x.iter().zip(&state.x).map(|(a, b)| a - b).collect(),
                grad.iter().zip(&state.grad).map(|(a, b)| a - b).collect(),
            ));
        } else {
            let (s, y) = state.s_y.last_mut().expect("m > 0");
            for i in 0..n {
                s[i] = x[i] - state.x[i];
                y[i] = grad[i] - state.grad[i];
            }
        }
        state.s_y.rotate_right(1);

        state.x.copy_from_slice(x);
        state.grad.copy_from_slice(&grad);

        for (j, (s_j, y_j)) in state.s_y.iter().enumerate() {
            rho[j] = 1. / (dot(y_j, s_j) + cfg.epsd);
        }

        q.copy_from_slice(&grad);

        for (j, (s_j, y_j)) in state.s_y.iter().enumerate() {
            let alpha_j = rho[j] * dot(s_j, &q);
            alpha[j] = alpha_j;
            for i in 0..n {
                q[i] -= alpha_j * y_j[i];
            }
        }

        // see page 226 of Nocedal and Wright
        let (s_k, y_k) = &state.s_y[0];
        let gamma = dot(s_k, y_k) / (dot(y_k, y_k) + cfg.epsd);
        for i in 0..n {
            r[i] = gamma * q[i];
        }

        for (j, (s_j, y_j)) in state.s_y.iter().enumerate().rev() {
            let alpha_j = alpha[j];
            let beta = rho[j] * dot(y_j, &r);
            for i in 0..n {
                r[i] += s_j[i] * (alpha_j - beta);
            }
        }

        let t = line_search(cfg, &mut f, &state.x, &r, fx, &mut grad, x);

        if let Some(msg) = stop(Info {
            state,
            fx,
            r: &r,
            t,
            x,
        }) {
            return msg;
        }
    }
}

File: packages/core/src/engine/Lbfgs.ts
// This module implements an L-BFGS optimizer using techniques from the
// following three books:
//
// - _Convex Optimization_ by Boyd and Vandenberghe, 2009 edition
// - _Engineering Optimization_ by Rao, 2009 edition
// - _Numerical Optimization_ by Nocedal and Wright, 1999 edition

/**
 * The first parameter is the point at which to evaluate the objective function,
 * and the second parameter is an output parameter to hold the gradient at that
 * point. The function should store the gradient in that output parameter and
 * then return the objective value.
 */
export type Fn = (x: Float64Array, grad: Float64Array) => number;

/** Configuration options for L-BFGS. */
export interface Config {
  /**
   * The number of vector pairs to store for L-BFGS. See page 224 of Nocedal and
   * Wright.
   */
  m: number;

  /** Constant for the Armijo condition. See page 37 of Nocedal and Wright. */
  armijo: number;

  /** Constant for the Wolfe condition. See page 39 of Nocedal and Wright. */
  wolfe: number;

  /** The minimum interval size for line search. */
  minInterval: number;

  /** The maximum number of steps for line search. */
  maxSteps: number;

  /** A small positive constant to add to a denominator that might be zero. */
  epsd: number;
}

/**
 * All L-BFGS state that needs to be kept between iterations, other than the
 * current point.
 */
export interface State {
  /** The previous point. */
  x: Float64Array;

  /** The previous gradient. */
  grad: Float64Array;

  /** See page 224 of Nocedal and Wright. */
  s: Float64Array[];

  /** See page 224 of Nocedal and Wright. */
  y: Float64Array[];
}

const dot = (u: Float64Array, v: Float64Array): number => {
  let z = 0;
  for (let i = 0; i < u.length; i++) z += u[i] * v[i];
  return z;
};

/**
 * Return the line search step size, having set `x` to the new point.
 *
 * @param x0 the current point.
 * @param r the descent direction, preconditioned by L-BFGS.
 * @param fx0 the objective at `x0`.
 * @param grad the gradient at `x0`, and is then used as scratch space; don't
 *  depend on its value.
 * @param x the new point. */
const lineSearch = (
  cfg: Config,
  f: Fn,
  x0: Float64Array,
  r: Float64Array,
  fx0: number,
  grad: Float64Array,
  x: Float64Array,
): number => {
  const n = x.length;

  const dufAtx0 = -dot(r, grad);

  let a = 0;
  let b = Infinity;
  let t = 1;
  let j = 0;

  for (;;) {
    for (let i = 0; i < n; i++) x[i] = x0[i] - t * r[i];

    if (Math.abs(a - b) < cfg.minInterval || j > cfg.maxSteps) break;

    const fx = f(x, grad);
    const isArmijo = fx <= fx0 + cfg.armijo * t * dufAtx0;
    const isWolfe = -dot(r, grad) >= cfg.wolfe * dufAtx0; // weak Wolfe condition

    if (!isArmijo) b = t;
    else if (!isWolfe) a = t;
    else break; // found good interval

    if (b < Infinity) t = (a + b) / 2; // already found Armijo
    else t = 2 * a; // did not find Armijo

    j++;
  }

  return t;
};

/**
 * Perform the first step of L-BFGS at point `x`, updating it and returning the
 * initial `State`.
 */
export const firstStep = (cfg: Config, f: Fn, x: Float64Array): State => {
  const n = x.length;
  const x0 = new Float64Array(x);

  const grad = new Float64Array(n);
  const fx = f(x, grad);

  const r = new Float64Array(grad);
  lineSearch(cfg, f, x0, r, fx, grad, x);

  return { x: x0, grad: r, s: [], y: [] };
};

/** Information after a step of L-BFGS. */
export interface Info {
  /** Data about previous steps. */
  state: State;

  /** The objective value at the current point. */
  fx: number;

  /** The preconditioned descent direction. */
  r: Float64Array;

  /** The current point. */
  x: Float64Array;

  /** The line search step size. */
  t: number;
}

/** Perform L-BFGS steps on `x`, giving `stop`'s first non-`undefined` value. */
export const stepUntil = <T>(
  cfg: Config,
  f: Fn,
  x: Float64Array,
  state: State,
  stop: (info: Info) => T | undefined,
): T => {
  const n = x.length;
  const grad = new Float64Array(n);

  const rho = new Float64Array(cfg.m);
  const alpha = new Float64Array(cfg.m);
  const q = new Float64Array(n);
  const r = new Float64Array(n);

  for (;;) {
    const fx = f(x, grad);

    let m = state.s.length;
    if (m < cfg.m) {
      m++;

      const s = new Float64Array(n);
      for (let i = 0; i < n; i++) s[i] = x[i] - state.x[i];
      state.s.push(s);

      const y = new Float64Array(n);
      for (let i = 0; i < n; i++) y[i] = grad[i] - state.grad[i];
      state.y.push(y);
    } else {
      const s = state.s[m - 1];
      const y = state.y[m - 1];
      for (let i = 0; i < n; i++) {
        s[i] = x[i] - state.x[i];
        y[i] = grad[i] - state.grad[i];
      }
    }
    state.s.unshift(state.s.pop()!);
    state.y.unshift(state.y.pop()!);

    state.x.set(x);
    state.grad.set(grad);

    for (let j = 0; j < m; j++)
      rho[j] = 1 / (dot(state.y[j], state.s[j]) + cfg.epsd);

    q.set(grad);

    for (let j = 0; j < m; j++) {
      const s_j = state.s[j];
      const y_j = state.y[j];
      const alpha_j = rho[j] * dot(s_j, q);
      alpha[j] = alpha_j;
      for (let i = 0; i < n; i++) q[i] -= alpha_j * y_j[i];
    }

    // see page 226 of Nocedal and Wright
    const s_k = state.s[0];
    const y_k = state.y[0];
    const gamma = dot(s_k, y_k) / (dot(y_k, y_k) + cfg.epsd);
    for (let i = 0; i < n; i++) r[i] = gamma * q[i];

    for (let j = m - 1; j >= 0; j--) {
      const s_j = state.s[j];
      const y_j = state.y[j];
      const alpha_j = alpha[j];
      const beta = rho[j] * dot(y_j, r);
      for (let i = 0; i < n; i++) r[i] += s_j[i] * (alpha_j - beta);
    }

    const t = lineSearch(cfg, f, state.x, r, fx, grad, x);

    const msg = stop({ state, fx, r, t, x });
    if (msg !== undefined) return msg;
  }
};
